update topic/graph-match/general-match (#10)
OI-wiki/i18n: docs/topic/graph-matching/graph-match.md

author: accelsao

**Matching** or **independent set of edges** is a set with no common edges in a graph. Finding a match in a bipartite graph is equivalent to the network flow problem.

> Graph matching algorithm is a commonly used algorithm in informatics competitions. It can be generally divided into maximum matching and maximum weight matching. We will first introduce the bipartite graph, and then propose solutions for general graph.

Reviewer: 这里信息学竞赛直接 OI 就行了，基本上整个 wiki 都在用这个

> Graph matching algorithm is a commonly used algorithm in informatics competitions. It can be generally divided into maximum matching and maximum weight matching. We will first introduce the bipartite graph, and then propose solutions for general graph.

Reviewer: ```suggestion
Graph matching algorithm are commonly used in informatics competitions. It can be generally divided into maximum matching and maximum weight matching. We will first introduce the bipartite graph, and then propose solutions for general graphs.
```

# Matching

> In graph theory, suppose the graph $G=(V,E)$ , where $V$ is vertex point set and $E$ is the edge set.

Reviewer: 统一直接用 vertex？:thinking:

A set of edges that have no common vertices in pairs $(M(M\in E))$ is called the **matching** of this graph.

Define the size of the match as the number of edges $|M|$ , and the largest number of edges $M$ is the **maximum match**.

> When the graph is weighted, the maximum sum of edge weight is **maximum weight matching**.

Reviewer: edges' weights

Reviewer: 另外这里可能还是要写成 the match with (max sum of weights) is  (max weight matching)？

> Maximal matching: No more matching edges can be added. It may not be the maximum match.
> Maximum matching: The match with the maximum number of matches.
> Perfect matching/Complete matching/1-factor: All vertices belong to the matching, and also meet the maximum match.
> Near-perfect matching: The number of vertices in the graph is odd, and there is exactly one vertex that is not in the match. The graph after removing this vertex is called factor-critical graph.

 **Maximal matching** ![graph-match-1](./images/graph-match-1.png) **maximum matching** ![graph-match-2](./images/graph-match-2.png)

# Bipartite graph matching

The matching on a bipartite graph is called bipartite matching. ![graph-match-3](./images/graph-match-3.png)

Let $G$ be a bipartite graph. If there are no common vertices in any two edges in the subgraph $M$ of $G$ , then $M$ is called a match of the bipartite graph $G$ , and the number of edges in $M$ is the number of matches.

## Complete match

Let $G=<V_1, V_2, E>$ be a bipartite graph, $|V_1| \leq |V_2|$ , $M$ is a maximum matching in $G$ , and $|M|=2|V_1|$ , then $M$ is called a complete match from $V_1$ to $V_2$ .

## Hall's Theorem

> Let bipartite graph $G=<V_1, V_2, E> , |V_1| \leq |V_2|$ , then there is a complete match between $V_1$ and $V_2$ in $G$ if and only for any $S \ subset V_1$ , both have $|S|\leq|N(S)|$ , where $N(S)=\Cup_{v_i \in S}{N(V_i)}$ is the neighborhood of $S$ .

Reviewer: leq 那里处理同上

Reviewer: 这里 both 有点迷惑诶 :thinking: 
指的是哪两个呢（纯提问）

另外邻域建议加个 wikipedia 的链接来更清楚指向？

Author: both应该说的是 `for any $S \ subset V_1$  => $|S|\leq|N(S)|$` :eyes: 是有有点误导, 这版我直接去掉了

## Maximum matching

Finding maximum number of matching edges in bipartite graph is called the maximum matching problem.

# Algorithm

A basic problem in combinatorial optimization is to find **maximum matching**.

## Bipartite graph maximum matching

> In the unweighted bipartite graph, the [Hopcroft-Karp algorithm](https://en.wikipedia.org/wiki/Hopcroft%E2%80%93Karp_algorithm) can be used to solve the problem in $O(\sqrt{V}E)$.

Reviewer: 这里做的很好 :+1:
原文那样表述其实是有语法错误的
